File: day-09/part-1/badouralix.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"
)

type node struct {
	value int
	prev  *node
	next  *node
}

func newnode(value int) *node {
	n := node{value, nil, nil}
	n.prev = &n
	n.next = &n
	return &n
}

func (this *node) insert(value int) *node {
	n := node{value, this, this.next}
	this.next.prev = &n
	this.next = &n
	return &n
}

func (this *node) delete() *node {
	this.prev.next = this.next
	this.next.prev = this.prev
	return this.next
}

func run(s string) interface{} {
	split := strings.Split(s, " ")
	players, _ := strconv.Atoi(split[0])
	last, _ := strconv.Atoi(split[6])

	scores := make([]int, players)
	current := newnode(0)
	player := 1
	result := 0

	for m := 1; m <= last; m++ {
		if m%23 != 0 {
			current = current.next.insert(m)
		} else {
			culprit := current.prev.prev.prev.prev.prev.prev.prev
			scores[player] += m + culprit.value
			current = culprit.delete()

		}

		player = player + 1
		if player >= players {
			player = 0
		}
	}

	for _, score := range scores {
		if score > result {
			result = score
		}
	}

	return result
}

func main() {
	input, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}
	start := time.Now()
	res := run(string(input))
	fmt.Printf("_duration:%f\n", time.Now().Sub(start).Seconds()*1000)
	fmt.Println(res)
}

File: day-09/part-1/badouralix.py
from runners.python import SubmissionPy


class BadouralixSubmission(SubmissionPy):

    def run(self, s):
        split = s.split(" ")
        players = int(split[0])
        last = int(split[6])
        # print(players, last)

        scores = [0] * players
        marbles = [0]
        pos = 0
        player = 1

        for m in range(last + 1):
            if m % 23 != 0:
                size = len(marbles)
                pos_l = (pos + 1) % size
                pos_r = (pos + 2) % size
                if pos_r == 0:
                    marbles = marbles + [m]
                    # pos = len(marbles) - 1
                else:
                    marbles = marbles[:pos_l + 1] + [m] + marbles[pos_r:]
                pos = pos_l + 1
            else:
                scores[player] += m
                pos = (pos - 7) % len(marbles)
                scores[player] += marbles[pos]
                marbles = marbles[:pos] + marbles[(pos + 1) % len(marbles):]

            player += 1
            player %= players

            if m < 1000:
                print(scores)

        return max(scores)
